workers/unity/Assets/Gdk/Core/Components/Option.cs

```diff
@@ -0,0 +1,120 @@
+using System;
+
+namespace Improbable.Gdk.Core
+{
+    public struct Option<T> : IEquatable<Option<T>>
+    {
+        private bool hasValue;
+        private T value;
+        
+        public Option(T newValue)
+        {
+            hasValue = false;
+            value = default(T);
+            Value = newValue;
+        }
+
+        public BlittableBool HasValue => hasValue;
+
+        public T Value
+        {
+            get
+            {
+                if (!hasValue)
+                {
+                    throw new InvalidOperationException("Called Value on empty Option.");
+                }
+
+                return value;
+            }
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentException("newValue is null.");
+                }
+
+                hasValue = true;
+                this.value = value;
+            }
+        }
+
+        public void Clear()
+        {
+            hasValue = false;
+            value = default(T);
+        }
+
+        public bool TryGetValue(out T value)
+        {
+            if (!hasValue)
+            {
+                value = default(T);
+                return false;
+            } 
+
+            value = this.value;
+            return true;
+        }
+
+        public T GetValueOrDefault(T defaultValue)
+        {
+            return hasValue ? value : defaultValue;
+        }
+
+        public override bool Equals(object other)
+        {
+            if (other is Option<T>)
+            {
+                return this.Equals((Option<T>)other);
+            }
+
+            return false;
+        }
+
+        public bool Equals(Option<T> other)
+        {
+            if (!hasValue && !other.HasValue)
+            {
+                return true;
+            }
+
+            if (hasValue && other.HasValue)
+            {
+                return value.Equals(other.Value);
```

Comment:
This line boxes as well I think (Value: `T` -> `object`) since `T` can be a value type

Reply:
As far as I understand, this boxes depending on whether the type of `value` overrides the default `Object.Equals` method. This sounds like something that is outside of our control.